v3_async_wip/v3_async_wip/iothub_mqtt_client.py

```diff
@@ -153,19 +153,19 @@ async def shutdown(self) -> None:
         # TODO: this breaks when called twice. Build some protections.
         # TODO: is there an issue with cancellation here?
         await self.disconnect()
-        # Cancel the SAS token update alarm. Note that this is not a task, it's a threaded Alarm.
-        # No need to wait for the result.
-        if self._sastoken_update_alarm:
-            logger.debug("Cancelling SAS Token update alarm")
-            self._sastoken_update_alarm.cancel()
-        # Cancel and wait for the completion of the twin response task
-        logger.debug("Cancelling twin response listener")
-        self._twin_response_listener.cancel()
+        cancelled_tasks = []
+
+        logger.debug("Cancelling 'process_twin_responses' background task")
+        self._process_twin_responses_bg_task.cancel()
+        cancelled_tasks.append(self._process_twin_responses_bg_task)
+
+        if self._keep_credentials_fresh_bg_task:
+            logger.debug("Cancelling 'keep_credentials_fresh' background task")
+            self._keep_credentials_fresh_bg_task.cancel()
+            cancelled_tasks.append(self._keep_credentials_fresh_bg_task)
+
         # Wait for the cancellation to complete before returning
-        try:
-            await self._twin_response_listener
-        except asyncio.CancelledError:
-            pass
+        await asyncio.gather(*cancelled_tasks, return_exceptions=True)
```

Comment:
> await asyncio.gather(*cancelled_tasks, return_exceptions=True) [](http://example.com/codeflow?start=7&length=63)

suggest adding a timeout. The cancel should be almost guaranteed, but I'm not sure if it is.

Reply:
Timeouts aren't really a thing with coroutines or gather, and trying to implement one on top would result in additional complexity and potentially problematic states. As long as the coroutine isn't hanging on a synchronous function and doesn't catch the CancelledError without re-raising (which we can establish via testing), there's no way for cancel to fail.

Reply:
Timeouts aren't properly supported with async until Python 3.11 - the idea really is that coroutines are supposed to just be cancellable. The only reasons a cancel wouldn't work is if the CancelledError is suppressed, or it's hanging on a synchronous call.

We can test to ensure that CancelledError doesn't get suppressed. Hanging on a synchronous call could be possible I suppose if the customer is using custom SAS token injection via synchronous callback, but we could even get around that by running their callbacks in a separate thread. Although that all being said, if a synchronous call is hanging, it's eating the entire event loop anyway, and you wouldn't even be able to invoke .shutdown() in the first place. So I suppose that's a moot point entirely.

Reply:
I don't know if I would go so far as to say "aren't really a thing".  `asyncio.wait_for` is a thing, and `asyncio.wait` has a timeout parameter. 3.11 added `asyncio.timeout` as a context manager object,. 

Your comment about "hanging on a synchronous function" is really important here because shutdown is a "last resort" function when things are going wrong, like powering off your phone or a three-fingered-salute. You don't want it to hang and you don't want it to partially complete.

If these tasks haven't completed after, say, 10 seconds, you don't have to worry about problematic states. You are already in a problematic state. Your best bet is to return and get out of the way and hope things don't get worse.